File: go-backend-interview-prep/05_errors_and_testing/04_sentinel_errors_vs_typed_errors/main.go
package main

import (
	"errors"
	"fmt"
)

// ---- Sentinel errors ----

var (
	ErrNotFound     = errors.New("not found")
	ErrUnauthorized = errors.New("unauthorized")
)

// ---- Typed error ----

type RateLimitError struct {
	Limit      int
	RetryAfter int // seconds
}

func (e *RateLimitError) Error() string {
	return fmt.Sprintf("rate limited: max %d req/s, retry after %ds", e.Limit, e.RetryAfter)
}

// ---- Functions ----

func lookupUser(id int) (string, error) {
	switch id {
	case 0:
		return "", ErrNotFound
	case -1:
		return "", ErrUnauthorized
	case 99:
		return "", &RateLimitError{Limit: 100, RetryAfter: 30}
	default:
		return fmt.Sprintf("user_%d", id), nil
	}
}

func getProfile(id int) (string, error) {
	name, err := lookupUser(id)
	if err != nil {
		return "", fmt.Errorf("getProfile(%d): %w", id, err)
	}
	return name, nil
}

func main() {
	ids := []int{1, 0, -1, 99}

	for _, id := range ids {
		name, err := getProfile(id)
		if err != nil {
			fmt.Printf("id=%d  error: %v\n", id, err)

			// Sentinel check with errors.Is
			if errors.Is(err, ErrNotFound) {
				fmt.Println("  -> sentinel: not found")
			}
			if errors.Is(err, ErrUnauthorized) {
				fmt.Println("  -> sentinel: unauthorized")
			}

			// Typed check with errors.As
			var rlErr *RateLimitError
			if errors.As(err, &rlErr) {
				fmt.Printf("  -> typed: limit=%d, retry_after=%ds\n", rlErr.Limit, rlErr.RetryAfter)
			}
		} else {
			fmt.Printf("id=%d  name: %s\n", id, name)
		}
		fmt.Println()
	}
}

File: go-backend-interview-prep/05_errors_and_testing/04_sentinel_errors_vs_typed_errors/main.py
"""Sentinel errors vs typed errors -- Python equivalent of the Go example."""


# ---- Sentinel-style exceptions (no fields) ----

class NotFoundError(Exception):
    pass


class UnauthorizedError(Exception):
    pass


# ---- Typed exception (has fields) ----

class RateLimitError(Exception):
    def __init__(self, limit: int, retry_after: int):
        self.limit = limit
        self.retry_after = retry_after
        super().__init__(f"rate limited: max {limit} req/s, retry after {retry_after}s")


# ---- Functions ----

def lookup_user(user_id: int) -> str:
    if user_id == 0:
        raise NotFoundError("not found")
    if user_id == -1:
        raise UnauthorizedError("unauthorized")
    if user_id == 99:
        raise RateLimitError(limit=100, retry_after=30)
    return f"user_{user_id}"


def get_profile(user_id: int) -> str:
    try:
        return lookup_user(user_id)
    except Exception as e:
        raise RuntimeError(f"get_profile({user_id})") from e


def main() -> None:
    ids = [1, 0, -1, 99]

    for uid in ids:
        try:
            name = get_profile(uid)
            print(f"id={uid}  name: {name}")
        except Exception as e:
            print(f"id={uid}  error: {e}")
            cause = e.__cause__

            # Sentinel-style check with isinstance
            if isinstance(cause, NotFoundError):
                print("  -> sentinel: not found")
            elif isinstance(cause, UnauthorizedError):
                print("  -> sentinel: unauthorized")

            # Typed check -- extract fields
            elif isinstance(cause, RateLimitError):
                print(f"  -> typed: limit={cause.limit}, retry_after={cause.retry_after}s")

        print()


if __name__ == "__main__":
    main()
